inst/include/dust/dust.hpp

```diff
@@ -299,14 +388,20 @@ class Dust {
   }
 
   std::vector<size_t> resample(const std::vector<real_t>& weights) {
-    std::vector<size_t> idx(n_particles());
+    std::vector<size_t> index(n_particles());
+    resample(weights, index);
+    return index;
+  }
+
+  void resample(const std::vector<real_t>& weights,
+                std::vector<size_t>& index) {
```

Comment:
Two thoughts:
- Why have the two separate functions here? Is there a case where we call the second with a non-empty index? What does that do?
- Why pass index by reference and require it to be constructed first? Would it not be simpler for this function to make index and return it instead of void?

Reply:
Complicated answer here:

* the second version (existing already, pass by reference) is the one that would be used in the inner loop of the particle filter; for that we create the index once and just use it over and over here. We could rewrite this to take an iterator though we've typically reserved that for cases where we do partial writes
* the first version then just wraps that but is used in the R-exposed resample method